### PR TITLE
Introduce GetOptions and ListOptions to storage interface

### PR DESCRIPTION
**Pull request by jpbetz**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Makes it much easier to add `ResourceVersionMatch` param as proposed in https://github.com/kubernetes/kubernetes/pull/91505

I've split this out from https://github.com/kubernetes/kubernetes/pull/91505 per https://github.com/kubernetes/kubernetes/pull/91505#discussion_r432692276 since it is an independent change

**Special notes for your reviewer**:

https://github.com/kubernetes/kubernetes/pull/91505 is based on this

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/priority important-longterm

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 1]
```diff
@@ -1142,46 +1195,55 @@ func TestList(t *testing.T) {
...
-			if !reflect.DeepEqual(wantPod, getPod) {
-				t.Errorf("(%s): pod want=%#v, got=%#v", tt.name, wantPod, getPod)
+			if err != nil {
+				if !tt.expectError {
+					t.Fatalf("List failed: %v", err)
+				}
+				return
 			}
-		}
+			if (len(out.Continue) > 0) != tt.expectContinue {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:1224]: Please validate that tt.expectError == nil at this point

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 2]
```diff
@@ -184,26 +184,23 @@ type Interface interface {
...
+	WatchList(ctx context.Context, key string, opts ListOptions) (watch.Interface, error)
 
 	// Get unmarshals json found at key into objPtr. On a not found error, will either
-	// return a zero object of the requested type, or an error, depending on ignoreNotFound.
+	// return a zero object of the requested type, or an error, depending on 'opts.ignoreNotFound'.
 	// Treats empty responses and nil response nodes exactly like a not found error.
-	// The returned contents may be delayed, but it is guaranteed that they will
-	// be have at least 'resourceVersion'.
-	Get(ctx context.Context, key string, resourceVersion string, objPtr runtime.Object, ignoreNotFound bool) error
+	// The returned contents may be delayed.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:192]: I agree that the previous comment wasn't ideal, but now it's too vague imho - then may be delayed only for RV!="" and the minRV semantic is guaranteed. Can you please update the comment?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:192]: Same below

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 3]
```diff
@@ -243,3 +240,22 @@ type Interface interface {
 	// Count returns number of different entries under the key (generally being path prefix).
 	Count(key string) (int64, error)
 }
+
+// GetOptions provides the options that may be provided for storage get operations.
+type GetOptions struct {
+	// IgnoreNotFound determines what is returned if the requested object is not found. If true, a zero object is
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:246]: nit: can you please shorten the lines?

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:192]: Actually, maybe what we need is:

... delayed according to the semantic of GetOptions.ResourceVersion.

or sth like that

**Review by wojtek-t — COMMENTED** [review]
Some comments, but nothing big. Thanks a lot for this refactoring Joe!

**Comment by fedebongio** [comment]
Thank you @wojtek-t for taking a look!

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/dryrun_test.go [hunk 4]
```diff
@@ -69,7 +69,7 @@ func TestDryRunCreateDoesntCreate(t *testing.T) {
 		t.Fatalf("Failed to create new dry-run object: %v", err)
 	}
 
-	err = s.Get(context.Background(), "key", "", out, false)
+	err = s.Get(context.Background(), "key", storage.GetOptions{ResourceVersion: ""}, out)
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/dryrun_test.go:72]: I agree. removing everywhere.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 5]
```diff
@@ -322,15 +322,16 @@ func (e *Store) ListPredicate(ctx context.Context, p storage.SelectionPredicate,
 	p.Continue = options.Continue
 	list := e.NewListFunc()
 	qualifiedResource := e.qualifiedResourceFromContext(ctx)
+	storageOpts := storage.ListOptions{options.ResourceVersion, p}
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:325]: Yes, I'll make sure they are used everywhere.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 6]
```diff
@@ -1083,15 +1088,16 @@ func (lw *cacherListerWatcher) List(options metav1.ListOptions) (runtime.Object,
 		Continue: options.Continue,
 	}
 
-	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, "", pred, list); err != nil {
+	// TODO(jpbetz): Shouldn't we set ResourceVersion?
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1091]: Thanks. Removing comment.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go [hunk 7]
```diff
@@ -50,9 +50,7 @@ func (a APIObjectVersioner) UpdateList(obj runtime.Object, resourceVersion uint6
 		return err
 	}
 	versionString := ""
-	if resourceVersion != 0 {
-		versionString = strconv.FormatUint(resourceVersion, 10)
-	}
+	versionString = strconv.FormatUint(resourceVersion, 10)
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go:53]: Heh, I hadn't noticed the previous line. Fixing!

> Also - should we return an error if resourceVersion == 0 ?

It's certainly not valid to have RV=0 in a response from etcd or the watch cache. Sure, I'll add it.

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:192]: Thanks, looks like this got messed up when I split this PR out from https://github.com/kubernetes/kubernetes/pull/91505. I'll add "delayed according to the semantic of GetOptions.ResourceVersion" for now and then merge in some more detail in https://github.com/kubernetes/kubernetes/pull/91505.

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 3 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:246]: Sure, I'll keep them under 100char/line which seems to be the norm for this file.

**Comment by jpbetz** [comment]
Thanks for the review @wojtek-t. Feedback applied.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 8]
```diff
@@ -618,7 +665,7 @@ func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {
 				return
 			default:
 				ctx, _ := context.WithTimeout(context.Background(), 3*time.Second)
-				w, err := cacher.Watch(ctx, "pods/ns", "0", pred)
+				w, err := cacher.Watch(ctx, "pods/ns", storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything})
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:668]: Please use pred - this test was added to catch a leak in bookmarks, which are set in `pred`, but not now (in storage.Everything).

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 9]
```diff
@@ -243,3 +240,23 @@ type Interface interface {
...
+	// ResourceVersion provides a resource version constraint to apply to the list operation
+	// as a "not older than" constraint: the result contains data at least as new as the provided
+	// ResourceVersion. The newest available data is preferred, but any data not older than this
+	// ResourceVersion may be served.
+	ResourceVersion string
+}
+
+// ListOptions provides the options that may be provided for storage list operations.
+type ListOptions struct {
+	// ResourceVersion provides a resource version constraint to apply to the list operation.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:258]: Might be worth unifying this comment with the one for GetOptions.ResourceVersion.

**Review by wojtek-t — COMMENTED** [review]
Two more minor comments - once fixed, lgtm

/approve

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 8 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:668]: Good catch. Fixed.

**Comment by wojtek-t** [comment]
@jpbetz - LGTM, please just squash commits and I will tag it.

**Comment by jpbetz** [comment]
squashed

**Pull request merged**